Use conda-build 3, new compilers, and new pinnings

jakirkham opened the pull request:
Brings this feedstock up-to-date by making a number of overdue changes.

<!--
Thank you for pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

Fixes https://github.com/conda-forge/pigz-feedstock/pull/2
Fixes https://github.com/conda-forge/pigz-feedstock/pull/5
Fixes https://github.com/conda-forge/pigz-feedstock/pull/7
Fixes https://github.com/conda-forge/pigz-feedstock/pull/8

<!--
Please add any other relevant info below:
-->

jakirkham committed:
MNT: Re-rendered with conda-build 3.18.11, conda-smithy 3.6.1, and conda-forge-pinning 2019.11.01

jakirkham committed:
Drop VC features

These have long been deprecated in favor of the new compilers.

jakirkham committed:
Split `build` and `host`

Move all dependencies that are linked against into `host`. Leave all
command line tools in `build`.

jakirkham committed:
Move compilers to the top of `build`

jakirkham committed:
Use the `compiler` syntax from conda-build 3

jakirkham committed:
MNT: Re-rendered with conda-build 3.18.11, conda-smithy 3.6.1, and conda-forge-pinning 2019.11.01

Now that the new `compiler` syntax is in use, re-render the feedstock to
update the variants accordingly.

jakirkham committed:
Unpin `zlib`

As pinnings are now handled globally using conda-forge-pinning and/or
locally using `run_exports` in packages (like `zlib` in this case), drop
the explicit pin from the recipe since it is no longer needed.

jakirkham committed:
MNT: Re-rendered with conda-build 3.18.11, conda-smithy 3.6.1, and conda-forge-pinning 2019.11.01

Now that `zlib` is unpinned, re-render to update the variants with this
information pulled from conda-forge-pinning.

jakirkham committed:
Drop deprecated `conda inspect` commands

As the `conda inspect` commands are deprecated and `conda-build` 3
includes this information, drop these unneeded commands from the tests.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jakirkham committed:
Switch to GitHub tarball URL

As it appears the zlib.net URL no longer exists for this version, switch
to using the GitHub tarball URL.

jakirkham committed:
Override the C compiler in the Makefile

Seems it tries to default to `gcc`, which does not work with the new
compilers. So set `CC` when calling `make`.

jakirkham committed:
Quote paths

jakirkham committed:
Make sure directories exist to copy into

jakirkham committed:
Ensure `CC` is set on Windows

Apparently the m2w64-toolchain does not set `CC`. So set `CC` to same
value it would have been on Windows.

jakirkham committed:
Bump build number to 1

Rebuild package now that conda-build 3, new compilers, and new pinnings
are used.

jakirkham commented on recipe/build.sh:
```diff
@@ -11,12 +11,19 @@ CFLAGS="$CFLAGS -O3 -I$PREFIX/include"
+# Set `CC` on Windows where it is not set already.
+if [ -z "$CC" ]; then
+    export CC="gcc"
+fi
```
Had to manually set `CC` for the MSYS2 toolchain on Windows. Raised issue ( https://github.com/conda-forge/toolchain-feedstock/issues/59 ) to address this.

jakirkham commented on recipe/meta.yaml:
```diff
@@ -1,44 +1,32 @@
 source:
   fn: {{ name }}-{{ version }}.tar.gz
-  url: https://zlib.net/{{ name }}/{{ name }}-{{ version }}.tar.gz
+  url: https://github.com/madler/{{ name }}/archive/v{{ version }}.tar.gz
```
This version is no longer available from https://zlib.net so have switched to the maintainer's GitHub repo. Would upgrade to the next version however that uses `fsync`, which does not work on Windows. Please see issue ( https://github.com/madler/pigz/issues/54 ) for details. Given there are already a lot of changes and it is probably worthwhile for the maintainers of this feedstock to approach that issue separately, have stuck to the current version so that we can get this feedstock up-to-date at least.

jakirkham commented:
This is now passing and ready for review, @epruesse! 😄

jakirkham commented:
@conda-forge-admin, please re-render.

jakirkham committed:
MNT: Re-rendered with conda-build 3.18.12, conda-smithy 3.6.10, and conda-forge-pinning 2020.03.06

jakirkham commented:
Going to go ahead and merge so we can get this recipe fixed up. Hope that is ok. If there are other issues that come up, happy to iterate on those. 😄
